### ISSUE TITLE
New home: f/flux and f/helm-operator docs

### ISSUE THREAD

**Issue by dholbach** [fluxcd/community#99]
In https://github.com/fluxcd/flux2/discussions/367 we sort of envisioned moving ALL docs to <https://fluxcd.io>.

While that's still a possibility it might make things more complicated and harder to understand for our users. We have been getting feedback that `v1` vs `v2` is a bit hard to grasp and having both set of docs on the same pages, might make the docs search hard to use.

Having distinctly differently looking sites for docs might not be so bad after all. I've been thinking about renaming `docs.fluxcd.io` (which hosts the docs for f/flux and f/helm-operator) to `legacy.fluxcd.io` to make things even clearer than just adding the *maintenance mode* warning at the top. This might just be hard to implement as we can't easily do redirects through readthedocs?

Any ideas about this? The goal of this enterprise would to make it even clearer to users which docs they're looking at, right now it's

- <https://fluxcd.io/docs> for newest Flux and
- <https://docs.fluxcd.io> for legacy Flux projects

Simple, right...?

**Comment by kingdonb** [comment]
I personally still find myself typing "toolki..." into the URL bar as I have muscle memory, and the docs in their new location seem much harder to simply rattle off and produce a reference URL off-the-cuff for any specific doc, than it was before.

It's a bit hard for me to express in plain words how much it frustrates me that docs moved under a subpath of fluxcd.io now.

It seems like the docs being front and center in the repo will make them more likely to get attention. Maybe it's just my imagination but they seem harder to reach in a number of ways.

I believe it would be forward-thinking to change the canonical location of Flux v1 docs links (to eg. `legacy-docs.fluxcd.io` with a temporary redirect for 2-6 months and a minor campaign to visit search engines periodically and make sure this change reaches them), so that in 6 months or so when the old `fluxcd/flux` repository is able to be permanently archived as maintenance is ended, this valuable real-estate `docs.fluxcd.io` is not permanently locked up by obsoleted legacy code and other historical but now irrelevant source materials. I'm not saying we should plan on moving the docs back out of the main domain and into a subdomain, but `docs.fluxcd.io` is one of the first places that people who are looking for docs will probably try when looking for Flux docs, and I think we should not keep hosting obsolete reference docs there in perpetuity.

I think the value of permalinks is not understated, but could be easily overstated when it comes to legacy and similarly EOL software, where the best thing for a new user is to reach a page that says instead "STOP: You're about to install a deprecated version, go here instead." This is unwelcoming for users of software in maintenance, so the temporary redirect should last as long as our maintenance commitment, then we start redirecting folks to an admonishment page directing them to upgrade.

I just don't know if I quite grasp the benefits we were supposed to realize from SEO and any other purpose by bringing the content all together completely under one domain, or how to quantify them against this seeming loss of function by reduced muscle memory. I do not think we need to keep the v1 docs in their current location as a permanent home. I would be just as happy if they don't get a new subdomain like `legacy-docs` and simply move under `fluxcd.io/legacy-docs/` instead.

In my ideal world, the slug `docs.fluxcd.io/%1` will always resolve to the same thing as `fluxcd.io/docs/%1`, (and both will be a reference that points at the most current docs).

I don't think we should give these valuable "funnel" locations over to historical documents for any longer than strictly needed.

**Comment by dholbach** [comment]
re muscle memory: Renaming and moving the docs were discussed in https://github.com/fluxcd/flux2/discussions/367 - a long time ago. If this frustrates folks so much, I would have hoped that people would have spoken up before. I hope that you'll find a way to get the URLs as fast now by using the url/search bar in your browser maybe... :man_shrugging: 

> It seems like the docs being front and center in the repo will make them more likely to get attention. Maybe it's just my imagination but they seem harder to reach in a number of ways.
Can you elaborate?

Some points regarding docs:
- docs.fluxcd.io is now on rtfd.io, which does things like: multiple versions, also deploying helm-op into `/projects/` for us
- readthedocs gives us https://docs.readthedocs.io/en/stable/user-defined-redirects.html in terms of redirects - AFAICT we'd be able to redirect `docs.` to `legacy.`
- the way I understand https://docs.readthedocs.io/en/stable/custom_domains.html#custom-domain-support we could deploy to two domains?

Maybe this would allow us to do something like:
1. Test deploy to two subdomains
2. Start deploying to `legacy.` and `docs.`
3. Test everything made it to both places
4. Redirect from `docs.` to `legacy.`
5. Wait some time
6. Reclaim `docs.` for other purposes...?


> In my ideal world, the slug docs.fluxcd.io/%1 will always resolve to the same thing as fluxcd.io/docs/%1, (and both will be a reference that points at the most current docs).

Right now that's not quite an option as we one is `v1` and the other `v2`, but maybe at some point in the future...?

**Comment by dholbach** [comment]
Regarding the search, we could potentially add another entry to https://github.com/algolia/docsearch-configs/commit/04fc07e3bde8337d753d067e3bee23d624e00f6c - so have two independent search things.

**Comment by kingdonb** [comment]
This is truly more of a gripe about the new doc structure than productive on-topic for the current issue...

Finding the right doc in the browser was easier before simply from a "number of keystrokes" perspective, as I could type the first few letters of the word "toolkit" and either press enter, instantly arriving at a landing page with huge Table of Contents sidebar, and search of relevant docs right where I can't miss it... or start completing to other docs URLs right there in the URL bar from those I have bookmarked and any I frequently read, immediately getting relevant results from the type ahead search.

That's working on the docs from one perspective, "how quickly can I navigate to the doc I need from the browser bar" - the first few keystrokes are unique to the docs site, "too..." and I'm already seeing docs pages in my browser bar. OK, so I type a couple of extra keystrokes and I find the document.

But then... from the other side, "how quickly I locate source markdown of said document and start editing from command line" has definitely regressed as well, to elaborate...

In the old scheme, I already had a clone of `flux2` and the docs were there in `docs/`. Type `cd fl[tab]do[tab]` and I'm there, at the hub of docs content. I wish we could invert the structure in `website` putting the content at the top of the repo not buried in some hideaway three subdirectories deep.

I can't just blast my way through the directory hierarchy with tab completion without stopping to resolve this, all of this just breaks up my flow and results in me becoming a bit flustered before I get started working on a doc.

I have to go to my `website` clone, which has `content/en/docs/` and it just feels like I'm about 6x as many keystrokes away from actually locating the doc I'm working on to open it up and start editing content. `cd we[tab]co[tab]` *swear at computer and bash loudly away at keyboard a couple of more times, as "`website`" matches also "`webui`" and "`content`" overlaps with "`config.toml`" too...* `n[tab][tab]do[tab]`... there, now we can start editing content.

I'm afraid this is too gripe-y and sounds really overly complaintative but I think if we had more contributors and solicit honest feedback from people who are working on docs every day we would find I'm not the only one swearing at their keyboard over these things. I'm not sure that's a good argument to make fixing them a priority, but it's something to be aware of before we make any more changes that have potential to make it better or worse.

One coping mechanism I found, I wound up making a symlink to ~/docs that reaches all the way down into `fluxcd/website/content/en/docs` so that it is more readily accessible from the home directory in my normal workflow. I should probably grow up and use VSCode which reduces this to an accordion hierarchy.

**Comment by dholbach** [comment]
Understood. Let's see what we can do to make editing the docs through CLI more user-friendly. Using the Web UI it became more friendly through the "edit this page" links I feel.

But that's all not what this issue is about really. Can we file a separate issue about this? I meant to use this one just for the new home for v1 docs.

**Comment by scottrigby** [comment]
I'd like to see a version select in the Flux docs at https://fluxcd.io/docs/. Examples:

# Kubernetes
Looks similar to our docs. I wonder is there a standard plugin for doc versions?
Note on the legacy versions there is a message at the top saying those docs are not maintained.

<img width="1435" alt="Screen Shot 2021-06-01 at 11 49 09 AM" src="https://user-images.githubusercontent.com/407675/120353073-63019900-c2cf-11eb-85c2-b32f7d27ca60.png"><img width="1437" alt="Screen Shot 2021-06-01 at 11 49 49 AM" src="https://user-images.githubusercontent.com/407675/120353170-7b71b380-c2cf-11eb-8add-4010437b4b3d.png">

# Helm
Helm does a similar thing. One nice thing about the legacy (2.x) helm docs is the warning message is pinned even when users scroll down. This is nice because if deep links have anchors, the warning is still visible.

<img width="1433" alt="Screen Shot 2021-06-01 at 11 51 01 AM" src="https://user-images.githubusercontent.com/407675/120353407-a5c37100-c2cf-11eb-89c6-8b1fcc97c55a.png">
<img width="1440" alt="Screen Shot 2021-06-01 at 11 52 49 AM" src="https://user-images.githubusercontent.com/407675/120353775-e622ef00-c2cf-11eb-8c1e-cba3661f31ea.png">

**Comment by dholbach** [comment]
On docs.fluxcd.io (v1 onl)y we have this version picker already. For a site containing v1 and v2 this is not really applicable as it'd sort of assume that across both sets of code there are always docs equivalents.

**Comment by scottrigby** [comment]
I'm not sure there needs to be equivalent doc pages for each version. For example the Helm page from `2.16.7` above just links back to the root of `3.2.4` docs when using the dropdown there. We would want to do the best mapping job we can, but if all else fails just fall back to the root welcome page?

**Comment by kingdonb** [comment]
Restating to not get lost on our internal Slack;

I think rather than reclaim the whole domain at once, we could add some selective redirects to the most popular Flux v1 docs pages according to search results, with a manually curated fuzzy mapping "old page" -> "most relevant new page" even where the structure of the old docs is totally different from the new docs

It's not totally clear for me whether that stuff happens down the road though, probably not until we have established new canonical URL homes recognized by search engines for everything on the old docs site... or whether by that point we have already lost and wasted the precious google juice

We I guess have no capacity for analytics on the old site now. It would be good to at least know which pages are still receiving the most traffic so we can judge whether there is any gain to be made by redirecting from the more popular landing points to some Flux v2 doc that can be on-topic. Maybe we can establish some current metrics anyway by putting a proxy host in front.

> For a site containing v1 and v2 this is not really applicable as it'd sort of assume that across both sets of code there are always docs equivalents.

While this is true sometimes, there are still very likely many Flux v1 docs pages (hopefully we have something relevant for 80% of Flux v1 docs visitors) that may simply have an equivalent piece of reference work in Flux v2. If those reference works are still attracting new readers then we should try to find a way to redirect them to the new version of Flux.

It should be possible to do this without making old reference pages completely inaccessible, but I think that a substantial portion of the traffic to the old docs would be better served by redirecting the old location at somewhere in the new docs... eventually, very carefully, even surgically. It is hard to make this argument without any analytics about the old site and it is questionable how we can gain analytics without disturbing any existing popular links.

**Comment by hiddeco** [comment]
Redirecting "certain pages" to the new website sounds like opening a new gate of confusion, especially for those that will be running Flux v1 a little longer on some forgotten cluster, while later trying to find any documentation to figure out how they e.g. tweak some configuration value.

I also feel this issue is losing the scope of the core thing it is trying to address _right now_: guiding users towards the proper documentation for the version they are using. – Anything related to contributing to documentation from the perspective of a maintainer has very little to do with this.

**Comment by dholbach** [comment]
Maybe we can treat this as an umbrella :open_umbrella: issue and file issues for more actionable things across `f/community` and `f/website`?

**Comment by dholbach** [comment]
Please check out: https://github.com/fluxcd/website/pull/310 - it's a minimal set of changes to

1. bring back Flux docs (not everything will work)
2. re-add the version picker from upstream docsy

It's just an experiment.

**Comment by dholbach** [comment]
Now that #fluxcd/website#310 is merged, I filed https://github.com/fluxcd/website/issues/327 to keep track of the remaining issues.

Regarding redirects from docs.fluxcd.io, here's what I did in an experiment:
- https://github.com/dholbach/empty-flux-docs generates HTML files that redirect for all link on <https://test-flux.rtfd.io> using a sphinx extension
- the repo has tags for all versions in https://test-flux.readthedocs.io/sitemap.xml
- in the readthedocs dashboard of `test-flux`, I cut over to this repository and ran all the builds
- all links I can see redirect over to fluxcd.io/legacy/flux now

I need feedback from @fluxcd/maintainers if 
- this approach looks promising
- the "emtpy repositories" should live under fluxcd

If that's fine, I'll 
- cut over the real flux (v1) docs too (creating tags for all the flux v1 releases)
- create a similar repo for the helm operator and cut it over too

**Comment by dholbach** [comment]
This should be up for https://test-helm-op.readthedocs.io/en/latest/ (in the form of https://github.com/dholbach/empty-helm-op-docs) as well now.

**Comment by dholbach** [comment]
@hiddeco and I agreed to use https://github.com/fluxcd/legacy-docs-redirects and https://github.com/fluxcd/helm-operator-docs-redirects for this. Thanks for the feedback - the road ahead for redirects is clear now.

**Comment by dholbach** [comment]
The vast majority of items on https://github.com/fluxcd/website/issues/327 have been cleared. Closing this one now.
